src/server.py

```diff
@@ -0,0 +1,149 @@
+# Copyright 2023 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Functions to interact with jenkins server."""
+
+import logging
+import random
+import time
+import typing
+from pathlib import Path
+
+import ops
+import requests
+from pydantic import BaseModel
+
+logger = logging.getLogger(__name__)
+
+JENKINS_WORKDIR = Path("/var/lib/jenkins")
+AGENT_JAR_PATH = Path(JENKINS_WORKDIR / "agent.jar")
+AGENT_READY_PATH = Path(JENKINS_WORKDIR / "agents/.ready")
+ENTRYSCRIPT_PATH = Path(JENKINS_WORKDIR / "entrypoint.sh")
+
+USER = "_daemon_"
+
+
+class Credentials(BaseModel):
+    """The credentials used to register to the Jenkins server.
+
+    Attrs:
+        address: The Jenkins server address to register to.
+        secret: The secret used to register agent.
+    """
+
+    address: str
+    secret: str
+
+
+class ServerBaseError(Exception):
+    """Represents errors with interacting with Jenkins server."""
+
+
+class AgentJarDownloadError(ServerBaseError):
+    """Represents an error downloading agent JAR executable."""
+
+
+def download_jenkins_agent(server_url: str, container: ops.Container) -> None:
+    """Download Jenkins agent JAR executable from server.
+
+    Args:
+        server_url: The Jenkins server URL address.
+        container: The agent workload container.
+
+    Raises:
+        AgentJarDownloadError: If an error occurred downloading the JAR executable.
+    """
+    try:
+        res = requests.get(f"{server_url}/jnlpJars/agent.jar", timeout=300)
+        res.raise_for_status()
+    except (requests.HTTPError, requests.Timeout, requests.ConnectionError) as exc:
+        logger.error("Failed to download agent JAR executable from server, %s", exc)
+        raise AgentJarDownloadError(
+            "Failed to download agent JAR executable from server."
+        ) from exc
+
+    container.push(path=AGENT_JAR_PATH, make_dirs=True, source=res.content, user=USER)
+
+
+def validate_credentials(
+    agent_name: str,
+    credentials: Credentials,
+    container: ops.Container,
+    add_random_delay: bool = False,
+) -> bool:
+    """Check if the credentials can be used to register to the server.
+
+    Args:
+        agent_name: The Jenkins agent name.
+        credentials: Server credentials required to register to Jenkins server.
+        container: The Jenkins agent workload container.
+        add_random_delay: Whether random delay should be added to prevent parallel registration on
+            server.
+
+    Returns:
+        True if credentials and agent_name pairs are valid, False otherwise.
+    """
+    # IMPORTANT: add random delay to prevent parallel execution.
+    if add_random_delay:
+        # It's okay to use random since it's not used for sensitive data.
+        time.sleep(random.random())  # nosec
```

Comment:
`random.random()` is between 0 and 1 second, is this too short in terms of delay?

Reply:
I've been testing it on different machines, this seemed fine on my machine, but I do agree that it should be adjusted (it doesn't quite work on Github actions runner) but I think I need a bit more time to experiment on what the magic number should be. Would you have any recommendations?

Reply:
Because what could happen is that this delay, if too big, might cause the pebble service launched by the agent that picked up the same credential to be terminated resulting in the error status.

Reply:
Gotcha, thanks for the explanation!